### ISSUE TITLE
Cannot compile string_view_test.cc due to constexpr errors

### ISSUE THREAD

**Issue by qzmfranklin** [abseil/abseil-cpp#63]
Hi,

Got a compilation issue here. Any help is much appreciated! Thanks in advance.

I will provide as much information as I can. Let me know if more is needed for people to debug.

OS: Ubuntu 16.04
CPU: x86_64
Compiler: clang-5.0.0 (the clang-based toolchain is fully installed under `/opt/clang/5.0.0` on the OS)
branch: master (720c017e30339fd1786ce4aac68bc8559736e53f)
Build using Bazel.

Tweaked the original BUILD.bazel files to build when included as source.

Here is how I produced the problem:

The sandbox command was:
```
(cd /home/zhongming/.cache/bazel/_bazel_zhongming/252308e7db9d6378926a9d84bb91921c/execroot/torrent && \
  exec env - \
    PWD=/proc/self/cwd \
  /opt/clang/5.0.0/bin/clang -U_FORTIFY_SOURCE '-D_FORTIFY_SOURCE=1' -fstack-protector -fcolor-diagnostics -Wall -fno-omit-frame-pointer -isystem/opt/clang/5.0.0/include/c++/v1 -nostdinc++ '-std=c++17' -MD -MF bazel-out/linux_clang-fastbuild/bin/third_party/abseil-cpp/_objs/strings_string_view_test/third_party/abseil-cpp/absl/strings/string_view_test.pic.d '-frandom-seed=bazel-out/linux_clang-fastbuild/bin/third_party/abseil-cpp/_objs/strings_string_view_test/third_party/abseil-cpp/absl/strings/string_view_test.pic.o' -fPIC -D__CLANG_SUPPORT_DYN_ANNOTATION__ -iquote . -iquote bazel-out/linux_clang-fastbuild/genfiles -iquote external/bazel_tools -iquote bazel-out/linux_clang-fastbuild/genfiles/external/bazel_tools -isystem third_party/abseil-cpp -isystem bazel-out/linux_clang-fastbuild/genfiles/third_party/abseil-cpp -isystem external/bazel_tools/tools/cpp/gcc3 -isystem third_party/gtest/googlemock -isystem bazel-out/linux_clang-fastbuild/genfiles/third_party/gtest/googlemock -isystem third_party/gtest/googlemock/include -isystem bazel-out/linux_clang-fastbuild/genfiles/third_party/gtest/googlemock/include -isystem third_party/gtest/googletest -isystem bazel-out/linux_clang-fastbuild/genfiles/third_party/gtest/googletest -isystem third_party/gtest/googletest/include -isystem bazel-out/linux_clang-fastbuild/genfiles/third_party/gtest/googletest/include -Wall -Wextra -Weverything -Wno-c++98-compat-pedantic -Wno-comma -Wno-conversion -Wno-covered-switch-default -Wno-deprecated -Wno-disabled-macro-expansion -Wno-documentation -Wno-documentation-unknown-command -Wno-double-promotion -Wno-exit-time-destructors -Wno-extra-semi -Wno-float-conversion -Wno-float-equal -Wno-format-nonliteral -Wno-gcc-compat -Wno-global-constructors -Wno-nested-anon-types -Wno-non-modular-include-in-module -Wno-old-style-cast -Wno-packed -Wno-padded -Wno-range-loop-analysis -Wno-reserved-id-macro -Wno-shorten-64-to-32 -Wno-sign-conversion -Wno-switch-enum -Wno-thread-safety-negative -Wno-undef -Wno-unknown-warning-option -Wno-unreachable-code -Wno-unused-macros -Wno-weak-vtables -Wbitfield-enum-conversion -Wbool-conversion -Wconstant-conversion -Wenum-conversion -Wint-conversion -Wliteral-conversion -Wnon-literal-null-conversion -Wnull-conversion -Wobjc-literal-conversion -Wstring-conversion -Wno-c99-extensions -Wno-missing-noreturn -Wno-missing-prototypes -Wno-null-conversion -Wno-shadow -Wno-shift-sign-overflow -Wno-sign-compare -Wno-unused-function -Wno-unused-member-function -Wno-unused-parameter -Wno-unused-private-field -Wno-unused-template -Wno-used-but-marked-unused -Wno-zero-as-null-pointer-constant -fexceptions -no-canonical-prefixes -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c third_party/abseil-cpp/absl/strings/string_view_test.cc -o bazel-out/linux_clang-fastbuild/bin/third_party/abseil-cpp/_objs/strings_string_view_test/third_party/abseil-cpp/absl/strings/string_view_test.pic.o)
```

The error was:
```
third_party/abseil-cpp/absl/strings/string_view_test.cc:743:21: warning: unused variable 'b' [-Wunused-variable]
  absl::string_view b(s1);
                    ^
third_party/abseil-cpp/absl/strings/string_view_test.cc:882:31: error: constexpr variable 'cstr' must be initialized by a constant expression
  constexpr absl::string_view cstr(nullptr);
                              ^~~~~~~~~~~~~
/opt/clang/5.0.0/include/c++/v1/__string:217:53: note: read of dereferenced null pointer is not allowed in a constant expression
    length(const char_type* __s)  _NOEXCEPT {return __builtin_strlen(__s);}
                                                    ^
/opt/clang/5.0.0/include/c++/v1/string_view:229:25: note: in call to 'length(0)'
                : __data(__s), __size(_Traits::length(__s)) {}
                                      ^
third_party/abseil-cpp/absl/strings/string_view_test.cc:882:31: note: in call to 'basic_string_view(0)'
  constexpr absl::string_view cstr(nullptr);
                              ^
third_party/abseil-cpp/absl/strings/string_view_test.cc:926:41: error: constexpr variable 'const_begin_nullptr' must be initialized by a constant expression
  constexpr absl::string_view::iterator const_begin_nullptr = cstr.begin();
                                        ^                     ~~~~~~~~~~~~
/opt/clang/5.0.0/include/c++/v1/string_view:239:51: note: initializer of 'cstr' is not a constant expression
        const_iterator cbegin() const _NOEXCEPT { return __data; }
                                                         ^
/opt/clang/5.0.0/include/c++/v1/string_view:233:51: note: in call to '&cstr->cbegin()'
        const_iterator begin()  const _NOEXCEPT { return cbegin(); }
                                                         ^
third_party/abseil-cpp/absl/strings/string_view_test.cc:926:68: note: in call to '&cstr->begin()'
  constexpr absl::string_view::iterator const_begin_nullptr = cstr.begin();
                                                                   ^
third_party/abseil-cpp/absl/strings/string_view_test.cc:882:31: note: declared here
  constexpr absl::string_view cstr(nullptr);
                              ^
third_party/abseil-cpp/absl/strings/string_view_test.cc:927:41: error: constexpr variable 'const_end_nullptr' must be initialized by a constant expression
  constexpr absl::string_view::iterator const_end_nullptr = cstr.end();
                                        ^                   ~~~~~~~~~~
/opt/clang/5.0.0/include/c++/v1/string_view:242:51: note: initializer of 'cstr' is not a constant expression
        const_iterator cend()   const _NOEXCEPT { return __data + __size; }
                                                         ^
/opt/clang/5.0.0/include/c++/v1/string_view:236:51: note: in call to '&cstr->cend()'
        const_iterator end()    const _NOEXCEPT { return cend(); }
                                                         ^
third_party/abseil-cpp/absl/strings/string_view_test.cc:927:66: note: in call to '&cstr->end()'
  constexpr absl::string_view::iterator const_end_nullptr = cstr.end();
                                                                 ^
third_party/abseil-cpp/absl/strings/string_view_test.cc:882:31: note: declared here
  constexpr absl::string_view cstr(nullptr);
                              ^
1 warning and 3 errors generated.
```

**Comment by tituswinters** [comment]
That's ... odd.  The char* c'tor for string_view should be constexpr, as should the underlying call to char_traits::length.
http://eel.is/c++draft/string.view.cons
http://eel.is/c++draft/char.traits.specializations.char

Seems likely to be a stdlib bug?  But still perhaps something that we need to work around.

Assigning to @zhangxy988 to investigate.

**Comment by zhangxy988** [comment]
First, note that `absl::string_view` is just an alias to `std::string_view` in your setting.
Second, the issue here is: `char_traits<char>::length(nullptr)` is currently undefined by the standard. libstdc++ considers it to be 0, but libc++ thinks it is invalid (out of range) and forbids nullptr in constexpr context. See https://godbolt.org/g/CUnpGe for a minimal reproduction of the issue.
Regardless of whether the standard will define this behavior or not, I'll need to work around this issue in string_view_test by conditional compilation. We also need to update our internal integration test to cover this setting. Please allow a few days before the fix gets in.
Thanks for reporting this issue to us!

**Comment by qzmfranklin** [comment]
Hi @zhangxy988 ,

Thanks!
Ah, I believe you are correct! I am using `libc++` instead of `libstdc++` and the issue you linked is exactly what the error was about. Thanks for replying so promptly!

Please feel free to let me know if any more information and testing is needed on me.

Best,

Zhongming

**Comment by zhangxy988** [comment]
This issue should be fixed by the latest commit 4132ce25956a91e1224e0f205b7f8c326304a995.
